### ISSUE TITLE
Reduce the number of requests needed for an incremental sync

### ISSUE THREAD

**Issue by githubbob42** [githubbob42/mingle2github2#5288]
**[Mingle Card: 5646](https://liquidframeworks.mingle.thoughtworks.com/projects/alpine_mobile/cards/5646)**
This card has been replaced by back office card #1395 
------------------------------------------------------

https://liquidframeworks.mingle.thoughtworks.com/projects/back\_office/cards/1395

1395 beta package id: 

```
04t1Y000001DxJNQA0
```

#1395 does not replace this, both are needed.

Rest endpoint test:
-------------------

At sfdc workbench memu on the top: Utilities→ Rest Explorer

do POST

/services/apexrest/FXQA/getSyncHints

{"whitelist":\[{"name":"FXQA\_\_Job\_\_c","lastsyncdate":"2018-02-17T16:15:44.366Z"},{"name":"FXQA\_\_Ticket\_\_c","lastsyncdate":"2018-02-17T16:15:44.367Z"}\]}

Note: 
------

This requires a managed package change. We will need to discuss this before writing the test plan. Also, we need to confirm that the PR doesn’t break for people who do not have the managed package change.

**Replace FXQA with FX5 if testing the production managed package**

**Use [https://time.is/UTC](https://time.is/UTC)  to get the current UTC time. You need that for lastsyncdate**

Description
-----------

Field Engineers sometimes have tenuous internet connections.  When this happens reducing the number of requests made can actually improve the overall experience of syncing.

Acceptance Criteria
-------------------

_This change should only affect incremental sync. It should greatly reduce the number of requests needed to perform the sync.  It should not interfere with the checksum process used by transactional data._

_Analysis_
----------

_We could add an endpoint which give hints for what the sync should do with each data type.  The apex code below should be safe for any data type.  It will be too lose improve sync for transactional types but will not overly constrain so the query will be safe._

```
   global class SyncResponseItem {
       public String name;
       public Boolean changed;
       public Boolean deleted;
       public SyncResponseItem(String name, Boolean changed, Boolean deleted){
           this.name = name;
           this.changed = changed;
           this.deleted = deleted;
       }
   }

   global class SyncRequestItem {
       public String name;
       public Datetime lastsyncdate;
   }

@HttpPost
   global static List<SyncResponseItem> getSyncHints(List<SyncRequestItem> whitelist){
       List<SyncResponseItem> results = new List<SyncResponseItem>();
       for(SyncRequestItem req: whitelist) {
           Map<String, Schema.SObjectField> fields = Schema.getGlobalDescribe().get(req.name).getDescribe().fields.getMap();
           String extraPredicates = '';
           // right now this only works on lookups, there isn't enough detail in these queries for transactional data
           if(fields.containsKey('altsync\_\_c'))
           {
               extraPredicates += 'AND AltSync\_\_c=true';
           }
           Boolean hasDeletedField = fields.containsKey('isdeleted');

           Datetime dt = req.lastsyncdate;
           String soql = String.format('SELECT ID FROM {0} WHERE SystemModStamp >= :dt {1} LIMIT 1', new String\[\]{req.name, extraPredicates});
           Boolean hasChanges = !Database.query(soql).IsEmpty();

           Boolean hasDeleted = hasDeletedField;
           if (hasDeletedField)
           {
               String dsoql = String.format('SELECT ID FROM {0} WHERE SystemModStamp >= :dt {1} AND IsDeleted = true LIMIT 1 ALL ROWS', new String\[\]{req.name, extraPredicates});
               hasDeleted = !Database.query(dsoql).IsEmpty();
           }

           if(hasChanges || hasDeleted)
               results.add(new SyncResponseItem(req.name, hasChanges, hasDeleted));
       }
       return results;
   }
```

This reduces an incremental sync in mobileteam from 112 requests down to 37.   It only reduces the 

Related Stories
---------------

Tasks
-----

{{ table query: SELECT Number, Name, Owner, 'Task Status' WHERE Type = Task and Story = THIS CARD ORDER BY Status }}

Defects
-------

{{ table query: SELECT Number, Name, Owner, 'Status' WHERE Type = Defect and 'Related Story' = THIS CARD }}

Test Plan
---------

Test on desktop and iPad. There is a managed package component to this as well, package PR 735

This PR only affects incremental syncs, so all tests are in relation to incremental sync.

1.  With the apex permission turned off, do an incremental sync and confirm the number of requests made in the dev tools.
2.  Turn on permission to the syncHints class.
3.  Do another incremental sync, there should be a 50%+ drop in the number of requests made when there are no changes to the back office data.

Test Result: Pass
-----------------

Installed beta package id: 04t1Y000001DxJNQA0 on my sandbox

Mobile PR: 2716

User: lanrefield@seven.com.s1   /  lastpass

Hardware: Dell XPS PC, IPAD Pro.

OS: Windows 10 & IOS 11

Browser: Chrome & Safari 

[https://sandbox.fieldfx.com/](https://sandbox.fieldfx.com/)

**Added Test Case:**   Reduce the number of requests needed for an incremental sync

[https://lfw.testrail.io/index.php?/cases/view/16150](https://lfw.testrail.io/index.php?/cases/view/16150)

With the apex permission turned off i got 113 requests when i added a new ticket with two ticket items to a job.

!Capture-bc25fe.png!{height: 523px; width: 990px;}

With the apex permission turned on i got 32 requests when i added a new ticket with two ticket items to a job.

!Capture2-0c67de.png!{height: 536px; width: 1000px;}

I was able to perform a smoke test on the PR on IPAD Pro running on IOS 11.

**Comment by githubbob42** [comment]
Natasha : 06/21/2018 08:04 CDT
server side tested; mobile side was not

**Comment by githubbob42** [comment]
Lanre : 07/09/2018 15:32 CDT
Waiting on MP

**Comment by githubbob42** [comment]
**Pull Request #2716**

**Comment by githubbob42** [comment]
Patty : 10/30/2018 08:46 CDT
Documented as Performance Improvement in 8/2/2018 Mobile release notes.
